[clang-format] Remove YAML hack to emit a BasedOnStyle comment

jasilvanus opened the pull request:
When serializing a formatting style to YAML, we were emitting a comment `# BasedOnStyle: <style>` if the serialized formatting style matches one of the known styles. This is useful, but mis-uses the YAML API.

An upcoming change to fix keys with special characters by quoting them breaks this, 
and will emit a non-comment **key** `'# BasedOnStyle': <style>` instead.
(https://github.com/llvm/llvm-project/pull/88763)

Thus, remove this hack.
Instead, we emit an ordinary key OrigBasedOnStyle that is ignored when reading back the data. 
An alternative would be to just remove it completely.

Ideally, we'd emit a proper comment, but our YAML API doesn't support that.

jasilvanus committed:
[clang-format] Remove YAML hack to emit a BasedOnStyle comment

When serializing a formatting style to YAML, we were emitting a
comment `# BasedOnStyle: <style>` if the serialized formatting
style matches one of the known styles. This is useful,
but mis-uses the YAML API.

An upcoming change to fix keys with special characters by quoting them
breaks this, and will emit a non-comment *key* `'# BasedOnStyle': <style>`
instead.
(https://github.com/llvm/llvm-project/pull/88763)

Thus, remove this hack.
Instead, we emit an ordinary key OrigBasedOnStyle that is ignored
when reading back the data. An alternative would be to just remove
it completely.

Ideally, we'd emit a proper comment, but our YAML API doesn't support
that.

mydeveloperday commented on clang/lib/Format/Format.cpp:
```diff
@@ -807,12 +807,18 @@ template <> struct MappingTraits<FormatStyle> {
+          // setting BasedOnStyle will override all other keys when importing,
+          // so we set a helper key that is ignored when importing.
+          // Ideally, we'd want a YAML comment here, but that's not supported.
+          IO.mapOptional("OrigBasedOnStyle", StyleName);
```
I think clang-format will baff if there is styles in the .clang-format file we don't understand, its very unforgiving of new styles

mydeveloperday replied:
Why is using # not a valid YAML comment?

jasilvanus replied:
> I think clang-format will baff if there is styles in the .clang-format file we don't understand, its very unforgiving of new styles

This will set a key that is never read just for reference, as the old comment did. 

> Why is using # not a valid YAML comment?

Using `#` in the textual yaml is a valid comment and perfectly fine. 
However, here we are using a function to set a key named `# BasedOnStyle`, which
just happened to emit a correct YAML comment because the key is not properly quoted/escaped.

This breaks when fixing key quoting.

mydeveloperday commented on clang/lib/Format/Format.cpp:
```diff
@@ -807,12 +807,18 @@ template <> struct MappingTraits<FormatStyle> {
         FormatStyle PredefinedStyle;
         if (getPredefinedStyle(StyleName, Style.Language, &PredefinedStyle) &&
             Style == PredefinedStyle) {
-          IO.mapOptional("# BasedOnStyle", StyleName);
```
is the a way to emit the a comment into the YAML at the correct indentation level?

jasilvanus replied:
I don't think our YAML writer supports comments at all.

HazardyKnusperkeks replied:
How hard would it be to add that?

jasilvanus replied:
I don't know, but I'd prefer to avoid supporting it if it isn't even needed. We can still add it if we want this back.

mydeveloperday reviewed (changes requested):
This doesn't feel correct to me...

jasilvanus commented:
> This doesn't feel correct to me...

I'd also be fine with just removing the comment.

HazardyKnusperkeks commented on clang/lib/Format/Format.cpp:
```diff
@@ -807,12 +807,18 @@ template <> struct MappingTraits<FormatStyle> {
       }
     } else {
+      StringRef OrigBasedOnStyle; // ignored
+      IO.mapOptional("OrigBasedOnStyle", OrigBasedOnStyle);
```
If we go that way, why read it back? This does not have any effect.

jasilvanus replied:
This allows the key being present in the input, avoiding an unknown key error. But now with all of it removed, this is no longer relevant.

HazardyKnusperkeks reviewed (commented):
The question is, is there any real benefit of putting the commented style in the output?

jasilvanus committed:
[clang-format] Remove dummy field

This may cause more confusion than helping.
If we really need a comment, we can add YAML
support for comments and add the comment back.

jasilvanus commented:
I've removed the dummy field and the comment now. This should prevent confusion.
I don't know whether anyone actually uses the emitted comment, there is no explanation,
and tests pass without it.

I suggest that if we later learn that there are valid uses that require the comment, we should add proper comment
support in the YAML writer, and add a test for it.

mydeveloperday commented:
I'm ok with removing it if @owenca  and @HazardyKnusperkeks  are.

HazardyKnusperkeks reviewed (approved):
I don't see a use case for that comment, and thus I see no problem in removing it.
